fix: add id for activitypub follows

Johann150 opened the pull request:
# What
`Follow` activities are rendered with a proper `id` so that when servers only respond with an accept with the id, it can be processed correctly.

The respective activity is provided at the new URL.
# Why
fix #8218

# Additional Info
together with #8635 this allows to follow users on Owncast 0.0.11 correctly (#8165)

Johann150 committed:
add id for activitypub follows

mei23 commented:
> together with https://github.com/misskey-dev/misskey/pull/8635 this allows to follow users on Owncast 0.0.11 correctly (https://github.com/misskey-dev/misskey/issues/8165)

Need a parser for `https://example.com/follows/<follower id>/<followee id>` ?

Johann150 committed:
fix lint

Johann150 commented:
I'm not sure where you mean this parser should be added? Misskey resolves everything via HTTP at the moment as far as I know (#8688), so that is also what I did for now: https://github.com/misskey-dev/misskey/pull/8689/files#diff-f6e4f1a116f4168da1998803a9ddb39d8f8a395aeced91cf67c4e42519807996R229

mei23 commented:
I think need to interpret the Accept payload returned remotely.
https://github.com/misskey-dev/misskey/blob/fee2878b98253f2510ae2bc5c746c058a043ef82/packages/backend/src/remote/activitypub/kernel/accept/follow.ts

#8688
`Resolver.resolve()` only gets the JSON of the AP, not the information required for local DB processing.

Johann150 commented:
The `Accept` payload will be turned into the AP object by this `Resolver.resolve`:
https://github.com/misskey-dev/misskey/blob/fee2878b98253f2510ae2bc5c746c058a043ef82/packages/backend/src/remote/activitypub/kernel/accept/index.ts#L16-L19
This will then fetch the AP JSON from the new route I added. Just as before, the AP JSON representation should contain all the necessary information. I'm not sure which information is missing.

mei23 commented:
Okay, I was worried about Undo and Reject processing, but maybe it's okay.

mei23 commented on packages/backend/src/server/activitypub.ts:
```diff
@@ -224,4 +225,24 @@ router.get('/likes/:like', async ctx => {
 });
 
+// follow
+router.get('/follows/:follower/:followee', async ctx => {
+	// This may be used before the follow is completed, so we do not
+	// check if the following exists.
+
+	const [follower, followee] = await Promise.all([
+		Users.findOneBy({ id: ctx.params.follower }),
+		Users.findOneBy({ id: ctx.params.followee }),
+	]);
+
+	if (follower == null || followee == null) {
+		ctx.status = 404;
+		return;
+	}
+
+	ctx.body = renderActivity(renderFollow(follower, followee));
+	ctx.set('Cache-Control', 'public, max-age=180');
+	setResponseType(ctx);
+});
```
うーん、インスタンスに保存するユーザーのパターンなら、どんなパターンでも作らされてしまうのが少し気になってる。

Johann150 replied:
Yes, it will be technically possible to create any pair of users stored in the database.

As I have said above already, being able to create this means little: Misskey uses follow requests to track whether a follow was requested by a local user. Before accepting a follow, Misskey checks that a matching follow request exists. If it does not exist, nothing happens.

Similarly, using this ID to create a fraudulent Activity should also not work because the actor that made the signature will not be the same as the `actor` of the `Follow` activity. This would only be a concern if the private key becomes public, which is in my understanding outside the threat model of Misskey.

To ensure this is not abused I think it does make sense to check that the follower is a local user and the followee is not a local user by adding those parameters to the queries.

If this is not what you meant, please be more specific.

mei23 replied:
まあ、そうは思うわ…

やっぱり、HTTP Resolver経由で取得するのが引っかかる。

Activity endpointは公開せずに /packages/backend/src/remote/activitypub/kernel/accept/follow.ts などでURLパースでDB解決するとかしたい。
Related #8688

Johann150 replied:
The resolve of the accepted object happens in `packages/backend/src/remote/activitypub/kernel/accept/index.ts`, and there a `Resolver` is used. Parsing the URL instead of making a HTTP request is the job of `DbResolver`. I am not sure if I can replace `Resolver` with a `DbResolver` there without breaking behaviour.

If it is possible, a new `getFollowFromApId` function could be added to `DbResolver` to be used instead, since misskey only seems to be able to accept `Follow` objects there anyway.

mei23 replied:
I was just thinking about it.

In the case of Undo, there can be Follow and Block, but since it is from local, it can be identified by URL Path.

Johann150 replied:
As I said before, I think this should be handled in the issue you linked, because in my opinion, there needs to be a way to automatically fetching the right type of object locally, as I wrote in <https://github.com/misskey-dev/misskey/issues/8688#issuecomment-1128811932>.
An additional complexity is that the kernel always uses ActivityPub objects, so use of `DbResolver` might be inappropriate because it returns database objects.

Johann150 committed:
fix: follower must be local, followee must be remote

Misskey will only use ActivityPub follow requests for users that are local
and are requesting to follow a remote user. This check is to ensure that
this endpoint can not be used by other services or instances.

Johann150 committed:
fix: missing import

mei23 commented:
It may be another issue, but should the path that stores Activity of ActivityPub be isolated to /activities/ or another path?

mei23 commented:
There is a possibility of conflict when implementing the Web or other protocols, and in other implementations we often see isolation in a dedicated path.

Johann150 commented:
It might have been a good design idea if it had been made at an earlier point. I think even besides it being a separate issue, such a change has the potential to (badly) break federation for servers that keep track of the existing IDs (for example: Misskey keeps track of the remote IDs of notes).

So in Misskey's case I think it is too late to change this.

mei23 commented:
Yes, most objects are too slow, but I'm wondering what about new Activities.

Johann150 commented:
Having different URL styles would require changes to `DbResolver.parseUri` which might be used in the future e.g. for #8688. So I do not want to create a mess of different URL formats.

mei23 commented:
Okay, there is such an idea.

However, `DbResolver.parseUri` does not assume two arguments, so it may be necessary to change it after all ...

Johann150 commented:
It would still create a mess of different URLs which is not something I would want to introduce.

mei23 commented:
How about `blocks`? I think it's the same pattern as `follows`.

mei23 commented:
> How about `blocks`? I think it's the same pattern as `follows`.

But unlike `follows`, ` blocks` can use an immutable DB ID.

Johann150 commented:
Does it matter for blocks? I don't think there is any case where local blocks are handled via ActivityPub, since they do not have to be `Accept`ed.

But I guess it would be best practice anyway...

Johann150 committed:
render block with id

Johann150 committed:
fix comment

syuilo commented:
👍
